genie-agent/src/main/java/com/netflix/genie/agent/execution/statemachine/actions/InitializeAction.java

```diff
@@ -35,13 +39,35 @@
 @Lazy
 class InitializeAction extends BaseStateAction implements StateAction.Initialize {
 
+    private final AgentRegistrationService agentRegistrationService;
+    private final ExecutionContext executionContext;
+
+    InitializeAction(
+        final AgentRegistrationService agentRegistrationService,
+        final ExecutionContext executionContext
+    ) {
+        this.agentRegistrationService = agentRegistrationService;
+        this.executionContext = executionContext;
+    }
+
     /**
      * {@inheritDoc}
      */
     @Override
     protected Events executeStateAction(final StateContext<States, Events> context) {
         log.info("Initializing...");
-        //TODO implement this action
+
+        final String agentId;
+        try {
+            agentId = agentRegistrationService.registerAgent();
+        } catch (final AgentRegistrationException e) {
+            throw new RuntimeException("Failed to obtain agent id", e);
```

Comment:
So this harkens back to my other comment about the exceptions. What is the value here of catching a checked exception only to rewrap in a runtime? You could use Guava Throwables.Propogate here too.

Reply:
So there's the service abstraction, the service does not know how it is going to be used.
So the checked exception is the way for the service to say "this failed" and give you some context (exception trace).
The service can also still throw runtime exceptions.
In this particular case, at the layer above, we wrap it in a Runtime, because that's the contract for state actions.
If you want, let's sit down and think through exception strategies, I'm not married to the idea that this is the best way, we can always change as we go.